MAINT: np.pad: Add helper functions for producing slices along axes

eric-wieser opened the pull request:
Follows on from #11011

eric-wieser committed:
MAINT: np.pad: Add helper functions for producing slices along axes

charris commented:
@eric-wieser What is the current status of this?

eric-wieser commented:
Needs a base-switch  from master and back again now that the other PR is merged, which will hide the first merged commit from the diff

eric-wieser commented on numpy/lib/arraypad.py:
```diff
@@ -771,8 +764,7 @@ def _pad_ref(arr, pad_amt, method, axis=-1):
     if 'odd' in method:
-        edge_slice2 = tuple(slice(None) if i != axis else slice(x - 1, x)
-                            for (i, x) in enumerate(arr.shape))
+        edge_slice2 = _end_slice(arr.shape, 1, axis=axis)
```
The alternative I was considering was `_axis_slice(arr.shape, lambda n: slice(n-1, n), axis=axis)`, which covers a few more uses, but I wasn't sure if that was clearer.

mhvk replied:
Not clearer as written, but I think it would be clearer if one were to just pass in the required slice itself. This is possible, since, as I noted above, it is not necessary to know the actual dimension; here, one would just pass in `slice(-1, None)`.

mhvk commented on numpy/lib/arraypad.py:
```diff
@@ -74,6 +74,18 @@ def _round_ifneeded(arr, dtype):
+                 for (i, x) in enumerate(shape))
+
+
+def _end_slice(shape, n, axis):
```
One doesn't actually need the `shape` for either `_begin_slice` (obvious, it is not used), or `_end_slice`: just do
```
def _end_slice(ndim, n, axis):
    return tuple(slice(None) if i != axis else slice(-n, None) for i in range(ndim))
```

eric-wieser replied:
`-n` fails when `n == 0`

mhvk replied:
Actually, thinking more, we do not need `ndim` either, since you only need a tuple up to `axis` - all the later dimensions are automatically taken to be `slice(None)`. So, it could be:
```
def _slice_at_axis(axis, slice):
    return (slice(None),) * axis + (slice,)
```
(This is substantially faster than the loop, at least for a randomly selected `axis=3`.)

mhvk replied:
Perhaps pass in `slice(arr.shape[axis] - padding, None)` in that case - at least it is immediately obvious what the code does.

eric-wieser replied:
I was trying to avoid writing `axis` twice, hence the lambda function above.

> This is substantially faster than the loop

I suspect this would be true if I kept my existing function names but changed their implementations too (I deliberately didn't do that in this PR, because it makes the diff clearer)

eric-wieser committed:
MAINT: np.pad: Generalize the helper function to be used in more places

This makes `_slice_first` almost a factor of two faster

eric-wieser commented:
Updated to build my helper functions out of the simpler helper function you propose (which is useful for a few other pieces of cleanup.

Also renamed from `_begin_slice` to `_slice_first`, which I think is clearer.

eric-wieser commented on numpy/lib/arraypad.py:
```diff
@@ -74,6 +74,25 @@ def _round_ifneeded(arr, dtype):
+    Construct a slice tuple the length of shape, with sl at the specified axis
+    """
+    slice_tup = (slice(None),)
+    return slice_tup * axis + (sl,) + slice_tup * (len(shape) - axis - 1)
```
Yes, the last half is redundant - but it produces the same slices as we did before, and I see little reason to change that.

eric-wieser commented on numpy/lib/arraypad.py:
```diff
@@ -479,11 +489,8 @@ def _append_mean(arr, pad_amt, num, axis=-1):
-    end = arr.shape[axis] - 1
     if num is not None:
-        mean_slice = tuple(
-            slice(None) if i != axis else slice(end, end - num, -1)
```
Note that this slice is the reverse of what it was before - but `min`, `max`, `mean`, and `med` all are order-invariant anyway

mhvk reviewed (approved):
I'm a bit surprised by the double slicing, but that was there already and I think this is all OK now, so approved! Maybe squash the two commits, though?

eric-wieser commented:
I think it's clearer as is with two commits - extracting helper functions from the existing code, then rewriting those helpers to be more efficient

eric-wieser commented:
Thanks for the feedback!
